File: scripts/convert-pretests-from-docx.py
import json
import re
import shutil
from pathlib import Path

from docx import Document
from docx.oxml.ns import qn
from docx.table import Table
from docx.text.paragraph import Paragraph


SOURCE_FILES = [
    Path(r"C:\Users\srava\Downloads\Pre-test\Grade 6 Pretest.docx"),
    Path(r"C:\Users\srava\Downloads\Pre-test\Grade 7 Pretest.docx"),
    Path(r"C:\Users\srava\Downloads\Pre-test\Grade 8 Pretest.docx"),
]

OUTPUT_DIR = Path("input/assessments")
ASSET_ROOT = Path("input/assets")
CATALOG_PATH = Path("input/assessment-catalog.json")

SUPERSCRIPT_MAP = str.maketrans({
    "0": "⁰",
    "1": "¹",
    "2": "²",
    "3": "³",
    "4": "⁴",
    "5": "⁵",
    "6": "⁶",
    "7": "⁷",
    "8": "⁸",
    "9": "⁹",
    "+": "⁺",
    "-": "⁻",
    "=": "⁼",
    "(": "⁽",
    ")": "⁾",
    "n": "ⁿ",
    "i": "ⁱ",
})

SUBSCRIPT_MAP = str.maketrans({
    "0": "₀",
    "1": "₁",
    "2": "₂",
    "3": "₃",
    "4": "₄",
    "5": "₅",
    "6": "₆",
    "7": "₇",
    "8": "₈",
    "9": "₉",
    "+": "₊",
    "-": "₋",
    "=": "₌",
    "(": "₍",
    ")": "₎",
})


def slugify(value):
    return re.sub(r"(^-|-$)", "", re.sub(r"[^a-z0-9]+", "-", value.lower())).strip("-")


def iter_blocks(document):
    body = document.element.body
    for child in body.iterchildren():
        if child.tag == qn("w:p"):
            yield Paragraph(child, document)
        elif child.tag == qn("w:tbl"):
            yield Table(child, document)


def paragraph_images(paragraph, document, asset_dir, slug, image_counter):
    images = []
    for blip in paragraph._p.xpath(".//a:blip"):
        embed = blip.get(qn("r:embed"))
        if not embed or embed not in document.part.related_parts:
            continue
        part = document.part.related_parts[embed]
        ext = Path(part.partname).suffix or ".png"
        image_counter[0] += 1
        filename = f"image{image_counter[0]}{ext}"
        output_path = asset_dir / filename
        output_path.write_bytes(part.blob)
        images.append(f"/input/assets/{slug}/{filename}")
    return images


def table_to_text(table):
    rows = []
    for row in table.rows:
        cells = [clean_inline_text(cell.text) for cell in row.cells]
        if any(cells):
            rows.append(" | ".join(cells))
    return " ".join(rows)


def table_images(table, document, asset_dir, slug, image_counter):
    images = []
    for blip in table._tbl.xpath(".//a:blip"):
        embed = blip.get(qn("r:embed"))
        if not embed or embed not in document.part.related_parts:
            continue
        part = document.part.related_parts[embed]
        ext = Path(part.partname).suffix or ".png"
        image_counter[0] += 1
        filename = f"image{image_counter[0]}{ext}"
        output_path = asset_dir / filename
        output_path.write_bytes(part.blob)
        images.append(f"/input/assets/{slug}/{filename}")
    return images


def normalize_option_text(value):
    return re.sub(r"^[A-Da-d]\)\s*", "", value.strip())


def parse_distractor(line):
    match = re.match(r"^([A-Da-d])\)\s*(.*?)(?:\s*\(Lesson:\s*(.*?)\))?$", line.strip())
    if not match:
        return None
    return {
        "id": match.group(1).lower(),
        "feedback": match.group(2).strip(),
        "lesson": (match.group(3) or "").strip() or None,
    }


def split_option_lines(text):
    text = text.strip()
    matches = list(re.finditer(r"(?m)([A-Da-d])\)\s*", text))
    if len(matches) <= 1:
        return [text] if text else []
    lines = []
    for index, match in enumerate(matches):
        end = matches[index + 1].start() if index + 1 < len(matches) else len(text)
        lines.append(text[match.start():end].strip())
    return lines


def clean_inline_text(value):
    value = str(value or "").replace("\u00a0", " ")
    value = value.replace("\r", " ").replace("\n", " ").replace("\t", " ")
    return re.sub(r"\s+", " ", value).strip()


def format_run_text(run):
    text = run.text or ""
    if run.font.superscript:
        return text.translate(SUPERSCRIPT_MAP)
    if run.font.subscript:
        return text.translate(SUBSCRIPT_MAP)
    return text


def paragraph_text(paragraph):
    return clean_inline_text("".join(format_run_text(run) for run in paragraph.runs))


def append_text(existing, addition):
    existing = clean_inline_text(existing)
    addition = clean_inline_text(addition)
    return clean_inline_text(f"{existing} {addition}" if existing else addition)


def new_question(number, assessment_key):
    return {
        "id": f"{assessment_key}-q{number:02d}",
        "type": "mcq",
        "number": number,
        "level": "",
        "topic": "",
        "standard": "",
        "question": "",
        "image": None,
        "images": [],
        "imageDescription": None,
        "options": [],
        "answer": "",
        "explanation": "",
        "distractors": {},
        "correctAnswerText": "",
    }


def add_images(question, images):
    if not question or not images:
        return
    question["images"].extend(images)
    if not question["image"]:
        question["image"] = images[0]


def parse_docx(path):
    document = Document(path)
    title = next((paragraph_text(p) for p in document.paragraphs if paragraph_text(p)), path.stem)
    slug = slugify(title)
    asset_dir = ASSET_ROOT / slug
    if asset_dir.exists():
        shutil.rmtree(asset_dir)
    asset_dir.mkdir(parents=True, exist_ok=True)

    image_counter = [0]
    questions = []
    current = None
    mode = None

    for block in iter_blocks(document):
        if isinstance(block, Table):
            table_text = table_to_text(block)
            images = table_images(block, document, asset_dir, slug, image_counter)
            if current:
                if table_text:
                    current["question"] = append_text(current["question"], table_text)
                add_images(current, images)
            continue

        text = paragraph_text(block)
        images = paragraph_images(block, document, asset_dir, slug, image_counter)
        if not text and images:
            add_images(current, images)
            continue
        if not text:
            continue
        if set(text.replace("—", "-")) <= {"-"}:
            continue
        if text == title:
            continue

        question_match = re.match(r"^Question\s+(\d+)(?:\s*\((.*?)\))?:\s*(.*)$", text, re.I)
        if question_match:
            if current:
                questions.append(current)
            current = new_question(int(question_match.group(1)), slug)
            current["level"] = (question_match.group(2) or "").strip()
            current["topic"] = current["level"] or "General"
            current["question"] = clean_inline_text(question_match.group(3))
            add_images(current, images)
            mode = "question"
            continue

        if text.startswith("Standard Code:"):
            if current and current.get("question"):
                questions.append(current)
                current = None
            mode = "standard"
            pending_standard = text.split(":", 1)[1].strip()
            continue

        if text.startswith("Bloom"):
            pending_level = text.split(":", 1)[1].strip() if ":" in text else ""
            continue

        if text.startswith("Question Number:"):
            if current and current.get("question"):
                questions.append(current)
            number = int(re.search(r"\d+", text).group(0))
            current = new_question(number, slug)
            current["standard"] = locals().get("pending_standard", "")
            current["level"] = locals().get("pending_level", "")
            current["topic"] = current["standard"] or current["level"] or "General"
            mode = None
            continue

        if text == "Question:":
            mode = "question"
            continue

        if text == "Distractor:":
            mode = "distractor"
            continue

        if not current:
            continue

        if text.startswith("Correct Answer:"):
            answer_text = clean_inline_text(text.split(":", 1)[1])
            answer_match = re.match(r"^([A-Da-d])\)\s*(.*)$", answer_text)
            if answer_match:
                current["answer"] = answer_match.group(1).lower()
                current["correctAnswerText"] = clean_inline_text(answer_match.group(2))
            else:
                current["correctAnswerText"] = answer_text
            mode = None
            continue

        if text.startswith("Explanation:"):
            current["explanation"] = clean_inline_text(text.split(":", 1)[1])
            mode = None
            continue

        if mode == "distractor":
            parsed = parse_distractor(text)
            if parsed:
                current["distractors"][parsed["id"]] = {
                    "feedback": parsed["feedback"],
                    "lesson": parsed["lesson"] or current["topic"] or "Review",
                }
            continue

        option_lines = split_option_lines(text)
        if option_lines and all(re.match(r"^[A-Da-d]\)", line) for line in option_lines):
            for line in option_lines:
                option_id = line[0].lower()
                option_image = images[0] if len(option_lines) == 1 and images else None
                current["options"].append({
                    "id": option_id,
                    "label": normalize_option_text(line),
                    "image": option_image,
                })
            if not (len(option_lines) == 1 and images):
                add_images(current, images)
            mode = "options"
            continue

        add_images(current, images)

        if mode == "question":
            current["question"] = append_text(current["question"], text)

    if current:
        questions.append(current)

    for question in questions:
        question["topic"] = question["topic"] or question["standard"] or question["level"] or "General"
        if not question["correctAnswerText"] and question["answer"]:
            option = next((item for item in question["options"] if item["id"] == question["answer"]), None)
            question["correctAnswerText"] = option["label"] if option else ""

    return {
        "assessment": {
            "key": slug,
            "title": title,
            "sourceDocument": path.name,
            "durationMinutes": 30,
            "tools": {
                "calculator": True,
                "scratchpad": True,
                "imageZoom": True,
                "eliminator": True,
            },
            "inputFormatVersion": "mvp-1",
            "questionSource": "Converted from Word document for MVP input folder",
            "instructions": [
                "Choose one answer for every question.",
                "Use the question grid to move between questions.",
                "Submit when you have answered all questions.",
            ],
        },
        "questions": questions,
    }


def main():
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    catalog = []

    for source in SOURCE_FILES:
        payload = parse_docx(source)
        slug = payload["assessment"]["key"]
        output_path = OUTPUT_DIR / f"{slug}.json"
        output_path.write_text(json.dumps(payload, indent=2, ensure_ascii=False), encoding="utf-8")
        catalog.append({
            "key": slug,
            "title": payload["assessment"]["title"],
            "sourceDocument": payload["assessment"]["sourceDocument"],
            "questionCount": len(payload["questions"]),
            "durationMinutes": payload["assessment"]["durationMinutes"],
            "path": f"input/assessments/{slug}.json",
        })
        print(f"Wrote {output_path} with {len(payload['questions'])} question(s).")

    CATALOG_PATH.write_text(json.dumps({"assessments": catalog}, indent=2, ensure_ascii=False), encoding="utf-8")
    print(f"Wrote {CATALOG_PATH} with {len(catalog)} assessment(s).")


if __name__ == "__main__":
    main()

File: scripts/convert-pretests-from-docx.js
const fs = require("fs");
const os = require("os");
const path = require("path");
const { execFileSync } = require("child_process");

const SOURCE_FILES = [
  "C:\\Users\\srava\\Downloads\\Pre-test\\Grade 6 Pretest.docx",
  "C:\\Users\\srava\\Downloads\\Pre-test\\Grade 7 Pretest.docx",
  "C:\\Users\\srava\\Downloads\\Pre-test\\Grade 8 Pretest.docx"
];

const OUTPUT_DIR = path.join("input", "assessments");
const ASSET_ROOT = path.join("input", "assets");
const CATALOG_PATH = path.join("input", "assessment-catalog.json");

const SUPERSCRIPT_MAP = {
  "0": "\u2070",
  "1": "\u00b9",
  "2": "\u00b2",
  "3": "\u00b3",
  "4": "\u2074",
  "5": "\u2075",
  "6": "\u2076",
  "7": "\u2077",
  "8": "\u2078",
  "9": "\u2079",
  "+": "\u207a",
  "-": "\u207b",
  "=": "\u207c",
  "(": "\u207d",
  ")": "\u207e",
  n: "\u207f",
  i: "\u2071"
};

const SUBSCRIPT_MAP = {
  "0": "\u2080",
  "1": "\u2081",
  "2": "\u2082",
  "3": "\u2083",
  "4": "\u2084",
  "5": "\u2085",
  "6": "\u2086",
  "7": "\u2087",
  "8": "\u2088",
  "9": "\u2089",
  "+": "\u208a",
  "-": "\u208b",
  "=": "\u208c",
  "(": "\u208d",
  ")": "\u208e"
};

function main() {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  const catalog = [];

  for (const source of SOURCE_FILES) {
    const payload = parseDocx(source);
    const slug = payload.assessment.key;
    const outputPath = path.join(OUTPUT_DIR, `${slug}.json`);
    fs.writeFileSync(outputPath, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
    catalog.push({
      key: slug,
      title: payload.assessment.title,
      sourceDocument: payload.assessment.sourceDocument,
      questionCount: payload.questions.length,
      durationMinutes: payload.assessment.durationMinutes,
      path: `input/assessments/${slug}.json`
    });
    console.log(`Wrote ${outputPath} with ${payload.questions.length} question(s).`);
  }

  fs.writeFileSync(CATALOG_PATH, `${JSON.stringify({ assessments: catalog }, null, 2)}\n`, "utf8");
  console.log(`Wrote ${CATALOG_PATH} with ${catalog.length} assessment(s).`);
}

function parseDocx(sourcePath) {
  if (!fs.existsSync(sourcePath)) {
    throw new Error(`DOCX file was not found: ${sourcePath}`);
  }

  const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), "assessment-docx-"));
  try {
    expandDocx(sourcePath, tempDir);
    const documentXml = readXml(path.join(tempDir, "word", "document.xml"));
    const relationships = readRelationships(path.join(tempDir, "word", "_rels", "document.xml.rels"));
    const title = getTitle(documentXml, path.parse(sourcePath).name);
    const slug = slugify(title);
    const assetDir = path.join(ASSET_ROOT, slug);

    fs.rmSync(assetDir, { recursive: true, force: true });
    fs.mkdirSync(assetDir, { recursive: true });

    const blocks = extractBodyBlocks(documentXml);
    const questions = [];
    let current = null;
    let mode = null;
    let pendingStandard = "";
    let pendingLevel = "";
    let imageCounter = 0;

    for (const block of blocks) {
      const text = block.type === "table" ? tableToText(block.xml) : paragraphText(block.xml);
      const images = extractImages(block.xml, relationships, tempDir, assetDir, slug, () => {
        imageCounter += 1;
        return imageCounter;
      });

      if (!text && images.length) {
        addImages(current, images);
        continue;
      }
      if (!text) continue;
      if (text.replace(/\u2014/g, "-").split("").every((char) => char === "-")) continue;
      if (text === title) continue;

      const questionMatch = text.match(/^Question\s+(\d+)(?:\s*\((.*?)\))?:\s*(.*)$/i);
      if (questionMatch) {
        if (current) questions.push(current);
        current = newQuestion(Number(questionMatch[1]), slug);
        current.level = cleanInlineText(questionMatch[2] || "");
        current.topic = current.level || "General";
        current.question = cleanInlineText(questionMatch[3] || "");
        addImages(current, images);
        mode = "question";
        continue;
      }

      if (text.startsWith("Standard Code:")) {
        if (current?.question) {
          questions.push(current);
          current = null;
        }
        pendingStandard = cleanInlineText(text.split(":").slice(1).join(":"));
        mode = "standard";
        continue;
      }

      if (text.startsWith("Bloom")) {
        pendingLevel = text.includes(":") ? cleanInlineText(text.split(":").slice(1).join(":")) : "";
        continue;
      }

      if (text.startsWith("Question Number:")) {
        if (current?.question) questions.push(current);
        const numberMatch = text.match(/\d+/);
        current = newQuestion(Number(numberMatch?.[0] || questions.length + 1), slug);
        current.standard = pendingStandard;
        current.level = pendingLevel;
        current.topic = current.standard || current.level || "General";
        mode = null;
        continue;
      }

      if (text === "Question:") {
        mode = "question";
        continue;
      }

      if (text === "Distractor:") {
        mode = "distractor";
        continue;
      }

      if (!current) continue;

      if (text.startsWith("Correct Answer:")) {
        const answerText = cleanInlineText(text.split(":").slice(1).join(":"));
        const answerMatch = answerText.match(/^([A-Da-d])\)\s*(.*)$/);
        if (answerMatch) {
          current.answer = answerMatch[1].toLowerCase();
          current.correctAnswerText = cleanInlineText(answerMatch[2]);
        } else {
          current.correctAnswerText = answerText;
        }
        mode = null;
        continue;
      }

      if (text.startsWith("Explanation:")) {
        current.explanation = cleanInlineText(text.split(":").slice(1).join(":"));
        mode = null;
        continue;
      }

      if (mode === "distractor") {
        const parsed = parseDistractor(text);
        if (parsed) {
          current.distractors[parsed.id] = {
            feedback: parsed.feedback,
            lesson: parsed.lesson || current.topic || "Review"
          };
        }
        continue;
      }

      const optionLines = splitOptionLines(text);
      if (optionLines.length && optionLines.every((line) => /^[A-Da-d]\)/.test(line))) {
        for (const line of optionLines) {
          const optionId = line[0].toLowerCase();
          current.options.push({
            id: optionId,
            label: normalizeOptionText(line),
            image: optionLines.length === 1 && images.length ? images[0] : null
          });
        }
        if (!(optionLines.length === 1 && images.length)) addImages(current, images);
        mode = "options";
        continue;
      }

      addImages(current, images);
      if (mode === "question") current.question = appendText(current.question, text);
    }

    if (current) questions.push(current);

    for (const question of questions) {
      question.topic = question.topic || question.standard || question.level || "General";
      if (!question.correctAnswerText && question.answer) {
        const option = question.options.find((item) => item.id === question.answer);
        question.correctAnswerText = option?.label || "";
      }
    }

    return {
      assessment: {
        key: slug,
        title,
        sourceDocument: path.basename(sourcePath),
        durationMinutes: 30,
        tools: {
          calculator: true,
          scratchpad: true,
          imageZoom: true,
          eliminator: true
        },
        inputFormatVersion: "mvp-1",
        questionSource: "Converted from Word document for MVP input folder",
        instructions: [
          "Choose one answer for every question.",
          "Use the question grid to move between questions.",
          "Submit when you have answered all questions."
        ]
      },
      questions
    };
  } finally {
    fs.rmSync(tempDir, { recursive: true, force: true });
  }
}

function expandDocx(sourcePath, destination) {
  const zipPath = path.join(path.dirname(destination), `${path.basename(destination)}.zip`);
  fs.copyFileSync(sourcePath, zipPath);
  execFileSync("powershell", [
    "-NoProfile",
    "-Command",
    "Expand-Archive -LiteralPath $args[0] -DestinationPath $args[1] -Force",
    zipPath,
    destination
  ], { stdio: "pipe" });
  fs.rmSync(zipPath, { force: true });
}

function extractBodyBlocks(xml) {
  const bodyMatch = xml.match(/<w:body[\s\S]*?>([\s\S]*?)<\/w:body>/);
  const body = bodyMatch?.[1] || "";
  const blocks = [];
  const blockPattern = /<w:(p|tbl)\b[\s\S]*?<\/w:\1>/g;
  let match;
  while ((match = blockPattern.exec(body))) {
    blocks.push({ type: match[1] === "tbl" ? "table" : "paragraph", xml: match[0] });
  }
  return blocks;
}

function paragraphText(xml) {
  const runs = xml.match(/<w:r\b[\s\S]*?<\/w:r>/g) || [];
  return cleanInlineText(runs.map(formatRunText).join(""));
}

function formatRunText(runXml) {
  let text = "";
  const textPattern = /<w:t\b[^>]*>([\s\S]*?)<\/w:t>|<w:tab\s*\/>|<w:br\b[^>]*\/>/g;
  let match;
  while ((match = textPattern.exec(runXml))) {
    if (match[1] != null) text += decodeXml(match[1]);
    else text += " ";
  }
  if (/<w:vertAlign\b[^>]*w:val="superscript"/.test(runXml)) return translateChars(text, SUPERSCRIPT_MAP);
  if (/<w:vertAlign\b[^>]*w:val="subscript"/.test(runXml)) return translateChars(text, SUBSCRIPT_MAP);
  return text;
}

function tableToText(xml) {
  const rows = xml.match(/<w:tr\b[\s\S]*?<\/w:tr>/g) || [];
  return cleanInlineText(rows.map((row) => {
    const cells = row.match(/<w:tc\b[\s\S]*?<\/w:tc>/g) || [];
    return cells.map(paragraphText).filter(Boolean).join(" | ");
  }).filter(Boolean).join(" "));
}

function extractImages(xml, relationships, tempDir, assetDir, slug, nextCounter) {
  const images = [];
  const ids = Array.from(xml.matchAll(/r:embed="([^"]+)"/g)).map((match) => match[1]);
  for (const id of ids) {
    const target = relationships[id];
    if (!target || !target.startsWith("media/")) continue;
    const source = path.join(tempDir, "word", target.replace(/\//g, path.sep));
    if (!fs.existsSync(source)) continue;
    const ext = path.extname(source) || ".png";
    const filename = `image${nextCounter()}${ext}`;
    fs.copyFileSync(source, path.join(assetDir, filename));
    images.push(`/input/assets/${slug}/${filename}`);
  }
  return images;
}

function readRelationships(relsPath) {
  const xml = readXml(relsPath);
  const relationships = {};
  const pattern = /<Relationship\b([^>]*)\/>/g;
  let match;
  while ((match = pattern.exec(xml))) {
    const attrs = parseAttributes(match[1]);
    if (attrs.Id && attrs.Target) relationships[attrs.Id] = attrs.Target;
  }
  return relationships;
}

function parseAttributes(value) {
  const attrs = {};
  const pattern = /([\w:.-]+)="([^"]*)"/g;
  let match;
  while ((match = pattern.exec(value))) attrs[match[1]] = decodeXml(match[2]);
  return attrs;
}

function readXml(filePath) {
  return fs.readFileSync(filePath, "utf8");
}

function getTitle(documentXml, fallback) {
  const firstText = extractBodyBlocks(documentXml)
    .filter((block) => block.type === "paragraph")
    .map((block) => paragraphText(block.xml))
    .find(Boolean);
  return firstText || fallback;
}

function newQuestion(number, assessmentKey) {
  return {
    id: `${assessmentKey}-q${String(number).padStart(2, "0")}`,
    type: "mcq",
    number,
    level: "",
    topic: "",
    standard: "",
    question: "",
    image: null,
    images: [],
    imageDescription: null,
    options: [],
    answer: "",
    explanation: "",
    distractors: {},
    correctAnswerText: ""
  };
}

function addImages(question, images) {
  if (!question || !images.length) return;
  question.images.push(...images);
  if (!question.image) question.image = images[0];
}

function parseDistractor(line) {
  const match = line.trim().match(/^([A-Da-d])\)\s*(.*?)(?:\s*\(Lesson:\s*(.*?)\))?$/);
  if (!match) return null;
  return {
    id: match[1].toLowerCase(),
    feedback: match[2].trim(),
    lesson: (match[3] || "").trim() || null
  };
}

function splitOptionLines(text) {
  const trimmed = text.trim();
  const matches = Array.from(trimmed.matchAll(/([A-Da-d])\)\s*/g));
  if (matches.length <= 1) return trimmed ? [trimmed] : [];
  return matches.map((match, index) => {
    const end = matches[index + 1]?.index ?? trimmed.length;
    return trimmed.slice(match.index, end).trim();
  });
}

function normalizeOptionText(value) {
  return value.trim().replace(/^[A-Da-d]\)\s*/, "");
}

function appendText(existing, addition) {
  const left = cleanInlineText(existing);
  const right = cleanInlineText(addition);
  return cleanInlineText(left ? `${left} ${right}` : right);
}

function cleanInlineText(value) {
  return String(value || "")
    .replace(/\u00a0/g, " ")
    .replace(/[\r\n\t]+/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function decodeXml(value) {
  return String(value || "")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&amp;/g, "&");
}

function translateChars(value, map) {
  return String(value || "").split("").map((char) => map[char] || char).join("");
}

function slugify(value) {
  return String(value || "")
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/(^-|-$)/g, "");
}

main();
